src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -2194,6 +2187,16 @@ private static async Task ExecuteResultWriteResponse(IResult? result, HttpContex
         await EnsureRequestResultNotNull(result).ExecuteAsync(httpContext);
     }
 
+    private static Task WriteJsonResponse(HttpResponse response, object? value)
+    {
+        // Call WriteAsJsonAsync() with the runtime type to serialize the runtime type rather than the declared type
+        // and avoid source generators issues.
+        // https://github.com/dotnet/aspnetcore/issues/43894
+        // https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-polymorphism
+        return HttpResponseJsonExtensions.WriteAsJsonAsync(response, value, value?.GetType() ?? typeof(object), default);
```

Comment:
Nit: this should minimize branching
```suggestion
        return HttpResponseJsonExtensions.WriteAsJsonAsync(response, value, value is null ? typeof(object) : value.GetType(), default);
```

https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcDMACR2DC2A3utmbjnEgGzYAqAngA4Cm2A4i8AEoCuAdsACWUFo1ZIAFAHsARgCsWAY2DYAbmAA2vFgEpsAXgB86rToD8AOk7BxLSfvPnswZi2kAzGQuXBdAblJyLFwaejcOLj5BETE3BG9FFVNtPUMTDVTsIQhsfl5NTWxnV1ZPRN99EBSday47B0C0AF8gA=